protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java

```diff
@@ -107,23 +110,38 @@ public T parse(InputStream stream) {
             }
           }
         }
+        CodedInputStream cis;
+        if (stream instanceof ReadableBuffer) {
```

Comment:
>  Copy from Composite Byte Buffer to single Heap byte buffer.

I unsettled with that. It could be fine, allocating a single 1 GB buffer is worrisome, as it requires continuous address space. If we are actually supporting 32 bit, then I think we need to shard. Avoiding a single allocation also allows us to work in a smaller amount of memory, since we can release buffers as protobuf parses.

So I think it's just "not that simple." If we limit the fast path to some maximum size that allows making progress for the common case without getting too burdened with the large message case. Without the size cap, then I think we have to optimize for large messages at the same time.